src/directives/async.ts

```diff
@@ -0,0 +1,60 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {directive, Directive, noChange, Part} from '../lit-html.js';
+
+interface PromisedValue<T = unknown> {
+  promise: Promise<T>;
+  resolved: boolean;
+  value?: T;
+}
+
+const promises = new WeakMap<Part, PromisedValue>();
+
+/**
+ * Renders a Promise to a Part when it resolves. `defaultContent` will be
+ * rendered until the Promise resolves.
+ */
+export const async = (promise: Promise<any>, defaultContent: any = noChange):
+    Directive<Part> => directive((part: Part) => {
+      let promisedValue = promises.get(part);
+
+      if (promisedValue !== undefined && promisedValue.promise === promise) {
+        // Set the default content if the Promise hasn't resolved *even if
+        // we're getting the same Promise* because the default content
+        // could have changed.
+        if (promisedValue.resolved && defaultContent !== noChange) {
+          part.setValue(defaultContent);
+        }
+        return;
+      }
+
+      // The following code only ever runs once per Promise instance
+      promisedValue = {promise, resolved: false};
+      promises.set(part, promisedValue);
+      if (defaultContent !== noChange) {
+        part.setValue(defaultContent);
+      }
+
+      Promise.resolve(promise).then((value: any) => {
+        const currentPromisedValue = promises.get(part);
+        if (currentPromisedValue === undefined ||
+            currentPromisedValue !== promisedValue) {
+          return;
+        }
+        promisedValue.value = value;
+        part.setValue(value);
```

Comment:
What if the value of the part has changed by now? `then` callback checks whether a new `async(promise)` was passed but it doesn't check for non-`async(promise)` value

Reply:
This isn't currently supported. There's really no way to tell that this directive has been removed - it has to be used somewhat statically.

Reply:
I though we could solve it by adding a new property to part, `readonly originalValue: any`, which would represent the value that gets set from the template regardless of whether it's a directive or something else entirely. That would allow the `async` directive to check whether it's still the active value in the case of a template like

```ts
html`${async(somePromise, 'temp')}`
```

but it would only work if `until` is used directly inside of the part, not in cases where it's wrapped in another directive, e.g.

```ts
html`${when(value instanceof Promise, () => async(value, 'temp'), () => value)}`
```

Reply:
I think this is something to be designed and solved for directives in general, not just async. Right now there are lots of sharp edges when directives aren't essentially "static" and top-level. Replacing directives can lead to other issues too.